src/interaction-handlers/sokoban.ts

```diff
@@ -0,0 +1,90 @@
+import { LanguageKeys } from '#lib/i18n/LanguageKeys';
+import {
+	buildGameControls,
+	buildSokobanGameFromResolvableLevel,
+	buildSokobanGameFromVisualLevel,
+	Direction,
+	encodeResolvableLevel
+} from '#lib/utilities/sokoban';
+import { ActionRowBuilder, ButtonBuilder } from '@discordjs/builders';
+import { InteractionHandler } from '@skyra/http-framework';
+import { getSupportedLanguageT } from '@skyra/http-framework-i18n';
+import { ButtonStyle, MessageFlags, type APIButtonComponentWithCustomId } from 'discord-api-types/v10';
+
+const Root = LanguageKeys.Commands.Sokoban;
+
+export class UserHandler extends InteractionHandler {
+	public async run(interaction: InteractionHandler.ButtonInteraction, [directionOrRetry, startTimestampOrEncodedLevel, moves]: Parameters) {
+		const t = getSupportedLanguageT(interaction);
+		// if the restart button is pressed, reset the current level
+		if (directionOrRetry === 'retry') {
+			const encodedLevel = startTimestampOrEncodedLevel!.replaceAll('-', '.');
+			const levelResult = buildSokobanGameFromResolvableLevel(encodedLevel);
+			if (levelResult.isErr()) {
+				return interaction.update({
+					content: t(LanguageKeys.Commands.Sokoban.SokobanInvalidComponent, { value: levelResult.unwrapErr() }),
+					flags: MessageFlags.Ephemeral
+				});
+			}
+			const level = levelResult.unwrap();
+
+			return interaction.update({
+				content: level.toString(),
+				components: buildGameControls(encodeResolvableLevel(level.components), level.checkPossibleMoves()),
+				flags: MessageFlags.Ephemeral
+			});
+		}
+
+		const startTimestamp = Number(startTimestampOrEncodedLevel);
+
+		// parse components from the emojis in the interaction's message
+		const gameBoardResult = buildSokobanGameFromVisualLevel(interaction.message.content);
+		if (gameBoardResult.isErr()) {
+			return interaction.update({
+				content: t(LanguageKeys.Commands.Sokoban.SokobanInvalidComponent, { value: gameBoardResult.unwrapErr() }),
+				components: [],
+				flags: MessageFlags.Ephemeral
+			});
+		}
+		const level = gameBoardResult.unwrap();
+
+		// execute the user's move
+		const pushedBoxOption = level.executeMove(directionOrRetry);
+		const updatedLevel = level.toString();
+
+		// check win condition, if met, send victory message
+		if (level.checkWinCondition()) {
+			return interaction.update({
+				content: `${t(Root.Victory, { seconds: ((Date.now() - startTimestamp) / 1000).toFixed(2), moves })}\n${updatedLevel}`,
```

Comment:
An issue I have with `.toFixed()` is that it always forces the decimals, so if you take `42` seconds exactly, this will make it `42.00`